### PR TITLE
APPS-650 Ensure parts are in the completed state when waiting on upload

### PR DESCRIPTION
**Pull request by mhrvol**
In case any part is not in successful state after upload, file will be re-uploaded.

I was not able to reproduce this exact issue. Note that this does not solve problem with timeouts (slow internet connection) && cancellations - in these cases, file will be still stuck in open state, as 'close file' is not called at all.

### CODE REVIEW

File: src/python/dxpy/bindings/dxfile_functions.py
Comment (mhrvol) [inline src/python/dxpy/bindings/dxfile_functions.py:570]: Whats happening here?

In case a part did not finish successfully - it is not in completed state, we will try to reupload the file. I chose 2 retries.

**Review by r-i-v-a — APPROVED** [review]

File: src/python/dxpy/bindings/__init__.py
Comment (emiloslavsky) [inline src/python/dxpy/bindings/__init__.py:660]: Is 60 seconds long enough to not trigger false re-uploads?

File: src/python/dxpy/bindings/dxfile_functions.py
Comment (emiloslavsky) [inline src/python/dxpy/bindings/dxfile_functions.py:578]: how does this work when uploading from a stding fife (e.g. `dx upload --path x -`)?

File: src/python/dxpy/bindings/__init__.py
Comment (mhrvol) [reply src/python/dxpy/bindings/__init__.py:660]: I agree. However I was never able to reproduce the issue where certain part is not closed or it is still "in progress of download". Everytime code goes to this function, these parts were already closed.

I was wondering whether we should just re-upload it again without even waiting, but it seems to me that it might be the case that the part would be finished in a while.

File: src/python/dxpy/bindings/__init__.py
Comment (mhrvol) [reply src/python/dxpy/bindings/__init__.py:669]: there are no parts for empty file. I thought of this more like a safety measure, in case there is some weird state where there are no parts.

File: src/python/dxpy/bindings/__init__.py
Comment (mhrvol) [reply src/python/dxpy/bindings/__init__.py:674]: I thought the same, I did not want to leave the possibility of the code failing due to key not existing. If this is not possible, I can remove this.

File: src/python/dxpy/bindings/dxfile_functions.py
Comment (mhrvol) [reply src/python/dxpy/bindings/dxfile_functions.py:578]: I tried `dx upload /dev/stdin` which worked. I guess this is the same thing, right?

File: src/python/dxpy/bindings/__init__.py
Comment (geetduggal) [reply src/python/dxpy/bindings/__init__.py:660]: I share @emiloslavsky's concern on this one.  Part sizes can get pretty big so I would suggest a significantly longer timeout coupled with perhaps a ceilinged exponential backoff.  My guess is it will be hard to get historical metrics at the part-granularity so we will have to do judgement calls here.  We want to guard against releasing this and having uploads fail _en masse_ for many users.

Comment (geetduggal) [reply src/python/dxpy/bindings/__init__.py:660]: (Actually, I can query to see if logs on our file closers can give an idea here)

File: src/python/dxpy/bindings/__init__.py
Comment (geetduggal) [inline src/python/dxpy/bindings/__init__.py:679]: Consider the case where is_uploaded is true but we happened to exceed the timeout

File: src/python/dxpy/bindings/__init__.py
Comment (geetduggal) [inline src/python/dxpy/bindings/__init__.py:682]: Consider having n=11 as opposed n=7 for this.  This an important timeout because it's a global limit on the upload file (1+2+4+8+... geometric series). Setting n=11 will result in this closed form sum for the geometric series: `((1-2^(11+1))/(1-2))/60` which is an hour and change.  That seems reasonable, but I feel like we need real data to base this on before landing it.  We may want to even expose this parameter for cases of really large files.  I will chat with the data team on this more.

File: src/python/dxpy/bindings/__init__.py
Comment (geetduggal) [inline src/python/dxpy/bindings/__init__.py:673]: (Minor: consider alternatively using the `any` procedure in Python`

File: src/python/dxpy/bindings/dxfile_functions.py
Comment (geetduggal) [inline src/python/dxpy/bindings/dxfile_functions.py:475]: What motivated this change?

**Review by geetduggal — COMMENTED** [review]
Nice job @mhrvol!  Some comments related mostly to timeouts.  I'll do some research on my end as well.  We should also define a clear test plan for this one.

File: src/python/dxpy/bindings/__init__.py
Comment (geetduggal) [reply src/python/dxpy/bindings/__init__.py:673]: Got it. Thanks!

File: src/python/dxpy/bindings/__init__.py
Comment (geetduggal) [reply src/python/dxpy/bindings/__init__.py:676]: Sorry for my misunderstanding.  Yes, a smaller timeout makes sense perhaps setting `timeout` to  something like `((1-2^(7+1))/(1-2))` so it will allow all 7 retries to occur.

**Review by geetduggal — APPROVED** [review]
LGTM modulo minor comments. Thanks!

**Comment by emiloslavsky** [comment]
@mhrvol - can this be merged?

**Comment by mhrvol** [comment]
> @mhrvol - can this be merged?

I am meeting with Kurt today, there's a weird cache issue.  @emiloslavsky

**Pull request merged**
